test/v3/runtime-core/apiWatch.spec.ts

```diff
@@ -531,4 +531,30 @@ describe('api: watch', () => {
 
     expect(data2.value).toMatchObject([1])
   })
+
+  // #498
+  it('watchEffect should not lead to infinite loop when accessing', async () => {
+    let dummy = 0
+
+    const state = reactive({
+      data: [],
+      watchVar: 123,
+    })
+
+    watchEffect(() => {
+      // accessing
+      state.watchVar
+      // setting
+      state.data = []
+      dummy += 1
+    })
+
+    expect(dummy).toBe(1)
+
+    state.data = []
```

Comment:
This shouldn't run the watcher since `state.data` was accessed, not set. A regular Vue `$watch` works that way.

After `watchEffect` I think it should be:

```javascript
expect(dummy).toBe(1)
await nextTick()
expect(dummy).toBe(1)
```

which will fail on `master` and passes on #786